docs/ics/ics-xxx-signed-messages.md

```diff
@@ -0,0 +1,150 @@
+
+* `encode(b : B)` = `p1 || bytes("Signed Cosmos SDK Message: \n") || l || b`, where
+  * `l`: little endian uint64 encoding of the length of `b`
+  * `p1`: prefix to distinguish from normal transactions and other encoding cases
+* `encode(s : S, domainSeparator : B)` = `p2 || domainSeparator || amino(s)`, where
+  * `domainSeparator`: 32 byte encoding of the domain separator [see below](###DomainSeparator)
+  * `p2`: prefix to distinguish from normal transactions and other encoding cases
+
+> TODO: Figure out byte(s) prefix in the encoding to not have collisions with
+typical transaction signatures (JSON-encoded) and to distinguish the individual
+cases. This may require introducing prefixes to transactions.
+
+#### Assumptions
+
+`amino` is deterministic and seems to be injective -- at least when when
+concrete types are registered. Can we rely on this mechanism or do we need a
+custom true injective encoding?
+
+User-agents such Voyager, Metamask, and Ledger may provide human-readable and
+structured data, but this data must be consistent in terms of the field types
+and structure itself encoded via `amino`. This would require said applications
+to have an `amino` implementation. This may yet further warrant a simple custom
+encoding.
+
+> TODO: Do we need to implement a custom encoding scheme instead of or atop 
+amino similar or equal to EIP-712 `hashStruct`? As far as I know, this is mostly
+due to providing Soldity efficiencies and addressing limitations.
+
+### DomainSeparator
+
+Encoding structures can still lead to potential collisions and while this may be
+ok or even desired, it introduces a concern in that it could lead to two compatible
+signatures. The domain separator prevents collisions of otherwise identical
+structures. It is designed to unique per application use and is directly used in
+the signature encoding itself. The domain separator is also extensible where the
+protocol and application designer may introduce or omit fields to their liking,
+but we will provide a typical structure that can be used for proper separation
+of concerns:
+
+```golang
+type DomainSeparator struct {
+    name    string  // A user readable name of the signing origin or application.
+    chainID string  // The corresponding Cosmos chain identifier.
+    version uint16  // Version of the domain separator. A single major version should suffice.
+    salt    []byte  // Random data to further provide disambiguation.
+}
+```
+
+Application designers may choose to omit or introduce additional fields to a
+domain separator. However, users should provided with the exact information for
+which they will be signing (i.e. a user should always know the chain ID they are
+signing for).
+
+Given the set of all domain separators, the encoding of the domain separator
+is as follows:
+
+* `encode(domainSeparator : B)` = `sha256(amino(domainSeparator))`
+
+## API
+
+We now formalize a standard set of APIs that providers must implement:
```

Comment:
What sort of API are we trying to specify - a REST API? A set of library functions?

I think we just need to pick a format for now and have a canonical implementation in the SDK; different libraries can implement the parts they need.

Reply:
My goal here is to provide an API spec, irrelevant to the underlying implementation. In either case, I will update it to be more alined to a spec rather than a direct API 👍